File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/LibraryCachePersistence.kt
package com.bockmedia.console.domain.model

import android.content.Context
import com.bockmedia.console.data.api.bockJson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import java.io.File

/** Disk snapshot so the Library tab paints instantly on cold start. */
object LibraryCachePersistence {
    private const val FILE_NAME = "library_cache.json"
    private const val MAX_AGE_MS = 24 * 60 * 60 * 1000L
    /** Bump when bucket shape changes so stale playlist-only snapshots are discarded. */
    private const val CACHE_VERSION = 3

    @Serializable
    private data class PlayTargetDto(
        val type: String,
        val id: String? = null,
        val name: String? = null,
        val artist: String? = null,
        val path: String? = null,
        val title: String? = null,
        val displayTitle: String? = null,
        val seedKind: String? = null,
    )

    @Serializable
    private data class ItemDto(
        val id: String,
        val title: String,
        val subtitle: String,
        val kind: String,
        val artPath: String? = null,
        val playlistId: String? = null,
        val artistName: String? = null,
        val albumName: String? = null,
        val sortDate: Long = 0L,
        val playTarget: PlayTargetDto,
    )

    @Serializable
    private data class SnapshotDto(
        val cacheVersion: Int = CACHE_VERSION,
        val savedAtMs: Long,
        val playlists: List<ItemDto> = emptyList(),
        val artists: List<ItemDto> = emptyList(),
        val albums: List<ItemDto> = emptyList(),
        val offline: List<ItemDto> = emptyList(),
    )

    suspend fun save(context: Context, data: LibraryData) {
        if (data.playlists.isEmpty() && data.artists.isEmpty() && data.albums.isEmpty()) return
        withContext(Dispatchers.IO) {
            runCatching {
                val dto = SnapshotDto(
                    cacheVersion = CACHE_VERSION,
                    savedAtMs = System.currentTimeMillis(),
                    playlists = data.playlists.map { it.toDto() },
                    artists = data.artists.map { it.toDto() },
                    albums = data.albums.map { it.toDto() },
                    offline = data.offline.map { it.toDto() },
                )
                val json = bockJson.encodeToString(dto)
                val target = file(context)
                val tmp = File(target.parentFile, "$FILE_NAME.tmp")
                tmp.writeText(json)
                if (!tmp.renameTo(target)) {
                    target.writeText(json)
                    tmp.delete()
                }
            }
        }
    }

    suspend fun load(context: Context): LibraryData? = withContext(Dispatchers.IO) {
        runCatching {
            val f = file(context)
            if (!f.exists()) return@runCatching null
            val dto = bockJson.decodeFromString<SnapshotDto>(f.readText())
            if (dto.cacheVersion < CACHE_VERSION) return@runCatching null
            if (System.currentTimeMillis() - dto.savedAtMs > MAX_AGE_MS) return@runCatching null
            val library = LibraryData(
                playlists = dto.playlists.mapNotNull { it.toModel() },
                artists = dto.artists.mapNotNull { it.toModel() },
                albums = dto.albums.mapNotNull { it.toModel() },
                offline = dto.offline.mapNotNull { it.toModel() },
            )
            if (library.playlists.isEmpty() && library.artists.isEmpty() && library.albums.isEmpty()) {
                return@runCatching null
            }
            // Drop playlist-only snapshots saved before artists/albums were cached.
            if (library.artists.isEmpty() && library.albums.isEmpty() && library.playlists.isNotEmpty()) {
                return@runCatching null
            }
            library
        }.getOrNull()
    }

    fun clear(context: Context) {
        runCatching { file(context).delete() }
    }

    private fun file(context: Context) = File(context.filesDir, FILE_NAME)

    private fun LibraryItem.toDto() = ItemDto(
        id = id,
        title = title,
        subtitle = subtitle,
        kind = kind.name,
        artPath = artPath,
        playlistId = playlistId,
        artistName = artistName,
        albumName = albumName,
        sortDate = sortDate,
        playTarget = playTarget.toDto(),
    )

    private fun PlayTarget.toDto(): PlayTargetDto = when (this) {
        is PlayTarget.Playlist -> PlayTargetDto("playlist", id = id, name = name)
        is PlayTarget.Artist -> PlayTargetDto("artist", name = name)
        is PlayTarget.Album -> PlayTargetDto("album", name = name, artist = artist)
        is PlayTarget.Song -> PlayTargetDto("song", path = path, title = title)
        is PlayTarget.Radio -> PlayTargetDto(
            "radio",
            displayTitle = displayTitle,
            name = name,
            path = path,
            seedKind = seedKind.name,
        )
    }

    private fun ItemDto.toModel(): LibraryItem? {
        val itemKind = runCatching { LibraryItemKind.valueOf(kind) }.getOrNull() ?: return null
        val target = playTarget.toModel() ?: return null
        return LibraryItem(
            id = id,
            title = title,
            subtitle = subtitle,
            kind = itemKind,
            playTarget = target,
            artPath = artPath,
            playlistId = playlistId,
            artistName = artistName,
            albumName = albumName,
            sortDate = sortDate,
        )
    }

    private fun PlayTargetDto.toModel(): PlayTarget? = when (type) {
        "playlist" -> PlayTarget.Playlist(id.orEmpty(), name.orEmpty())
        "artist" -> name?.let { PlayTarget.Artist(it) }
        "album" -> name?.let { PlayTarget.Album(it, artist) }
        "song" -> path?.let { PlayTarget.Song(it, title ?: name ?: "Track") }
        "radio" -> {
            val seed = runCatching {
                PlayTarget.RadioSeedKind.valueOf(seedKind ?: "Artist")
            }.getOrDefault(PlayTarget.RadioSeedKind.Artist)
            PlayTarget.Radio(displayTitle ?: name.orEmpty(), seed, name.orEmpty(), path)
        }
        else -> null
    }
}

File: ios/BockMedia/Domain/LibraryCachePersistence.swift
import Foundation

enum LibraryCachePersistence {
    private static let fileName = "library_cache.json"
    private static let maxAgeMs: Int64 = 24 * 60 * 60 * 1000

    private struct PlayTargetDto: Codable {
        let type: String
        let id: String?
        let name: String?
        let artist: String?
        let path: String?
        let title: String?
        let displayTitle: String?
        let seedKind: String?
    }

    private struct ItemDto: Codable {
        let id: String
        let title: String
        let subtitle: String
        let kind: String
        let artPath: String?
        let playlistId: String?
        let artistName: String?
        let albumName: String?
        let sortDate: Int64
        let playTarget: PlayTargetDto
    }

    private struct SnapshotDto: Codable {
        let savedAtMs: Int64
        let playlists: [ItemDto]
        let artists: [ItemDto]
        let albums: [ItemDto]
        let offline: [ItemDto]
    }

    static func save(_ data: LibraryData) {
        if data.playlists.isEmpty && data.artists.isEmpty && data.albums.isEmpty { return }
        let dto = SnapshotDto(
            savedAtMs: Int64(Date().timeIntervalSince1970 * 1000),
            playlists: data.playlists.map(itemToDto),
            artists: data.artists.map(itemToDto),
            albums: data.albums.map(itemToDto),
            offline: data.offline.map(itemToDto)
        )
        guard let url = cacheURL(), let json = try? JSONEncoder().encode(dto) else { return }
        try? json.write(to: url, options: .atomic)
    }

    static func load() -> LibraryData? {
        guard let url = cacheURL(),
              let data = try? Data(contentsOf: url),
              let dto = try? JSONDecoder().decode(SnapshotDto.self, from: data) else { return nil }
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        if now - dto.savedAtMs > maxAgeMs { return nil }
        let library = LibraryData(
            playlists: dto.playlists.compactMap(dtoToItem),
            artists: dto.artists.compactMap(dtoToItem),
            albums: dto.albums.compactMap(dtoToItem),
            offline: dto.offline.compactMap(dtoToItem)
        )
        guard !library.playlists.isEmpty || !library.artists.isEmpty || !library.albums.isEmpty else { return nil }
        return library
    }

    static func clear() {
        guard let url = cacheURL() else { return }
        try? FileManager.default.removeItem(at: url)
    }

    private static func cacheURL() -> URL? {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?
            .appendingPathComponent(fileName)
    }

    private static func itemToDto(_ item: LibraryItem) -> ItemDto {
        ItemDto(
            id: item.id,
            title: item.title,
            subtitle: item.subtitle,
            kind: kindName(item.kind),
            artPath: item.artPath,
            playlistId: item.playlistId,
            artistName: item.artistName,
            albumName: item.albumName,
            sortDate: item.sortDate,
            playTarget: targetToDto(item.playTarget)
        )
    }

    private static func dtoToItem(_ dto: ItemDto) -> LibraryItem? {
        guard let kind = parseKind(dto.kind), let target = dtoToTarget(dto.playTarget) else { return nil }
        return LibraryItem(
            id: dto.id,
            title: dto.title,
            subtitle: dto.subtitle,
            kind: kind,
            playTarget: target,
            artPath: dto.artPath,
            playlistId: dto.playlistId,
            artistName: dto.artistName,
            albumName: dto.albumName,
            sortDate: dto.sortDate
        )
    }

    private static func kindName(_ kind: LibraryItemKind) -> String {
        switch kind {
        case .playlist: return "playlist"
        case .artist: return "artist"
        case .album: return "album"
        case .track: return "track"
        case .downloaded: return "downloaded"
        }
    }

    private static func parseKind(_ raw: String) -> LibraryItemKind? {
        switch raw.lowercased() {
        case "playlist": return .playlist
        case "artist": return .artist
        case "album": return .album
        case "track": return .track
        case "downloaded": return .downloaded
        default: return nil
        }
    }

    private static func targetToDto(_ target: PlayTarget) -> PlayTargetDto {
        switch target {
        case .playlist(let id, let name):
            return PlayTargetDto(type: "playlist", id: id, name: name, artist: nil, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .artist(let name):
            return PlayTargetDto(type: "artist", id: nil, name: name, artist: nil, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .album(let name, let artist):
            return PlayTargetDto(type: "album", id: nil, name: name, artist: artist, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .song(let path, let title):
            return PlayTargetDto(type: "song", id: nil, name: nil, artist: nil, path: path, title: title, displayTitle: nil, seedKind: nil)
        case .radio(let displayTitle, let seedKind, let name, let path):
            let seed = switch seedKind {
            case .artist: "Artist"
            case .song: "Song"
            case .genre: "Genre"
            }
            return PlayTargetDto(type: "radio", id: nil, name: name, artist: nil, path: path, title: nil, displayTitle: displayTitle, seedKind: seed)
        }
    }

    private static func dtoToTarget(_ dto: PlayTargetDto) -> PlayTarget? {
        switch dto.type {
        case "playlist":
            return .playlist(id: dto.id ?? "", name: dto.name ?? "")
        case "artist":
            guard let name = dto.name else { return nil }
            return .artist(name: name)
        case "album":
            guard let name = dto.name else { return nil }
            return .album(name: name, artist: dto.artist)
        case "song":
            guard let path = dto.path else { return nil }
            return .song(path: path, title: dto.title ?? dto.name ?? "Track")
        case "radio":
            let seed: PlayTarget.RadioSeedKind = switch dto.seedKind?.lowercased() {
            case "song": .song
            case "genre": .genre
            default: .artist
            }
            return .radio(displayTitle: dto.displayTitle ?? dto.name ?? "", seedKind: seed, name: dto.name ?? "", path: dto.path)
        default:
            return nil
        }
    }
}
